### PR TITLE
Feature/duplicated array keys

### PR DESCRIPTION
**Pull request by eeree**
As https://github.com/phpmd/phpmd/pull/350 stuck for some time and also implementation was a bit buggy (nested arrays recursion issue and boolean/null array keys checks), I've decided to fix what's left.

### CODE REVIEW

File: src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    }
+
+    /**
+     * Analyzes single array.
+     *
+     * @param ASTNode $node Array node.
+     * @return void
+     */
+    private function analyzeArray(ASTNode $node)
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php]: Very general name. How about ``checkForDuplicatedArrayKey``?

**Comment by ravage84** [comment]
@eeree if we ever meet, remember me to buy you a drink 😼

**Comment by ravage84** [comment]
Ignore the Stickler CI nags, created #487 for that...

File: src/test/resources/files/Rule/CleanCode/DuplicatedArrayKey/testRuleAppliesWhenKeyIsDeclaredInNonStandardWay.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            $baz => 'bar', // not applied - see above
+            $this->getDuplicatedName() => 'bar', // not applied - resolving variable depends on inheritance tree
+            self::$classStaticPropertyDuplicate => 'bar', // not applied - static property may be modified externally
+            self::$classPrivateStaticPropertyDuplicate, // not applied - see above
+            $this->classPrivatePropertyDuplicate => 'bar', // not applied - may be modified at any time
+            $this->classProtectedPropertyDuplicate => 'bar', // not applied - see above
+            $this->classPublicPropertyDuplicate => 'bar', // not applied - see above
+            global_function_duplicate() => 'bar', // not applied - may be different depending on namespace
+            GLOBAL_DUPLICATE => 'bar', // not applied - may be different depending on context
+            "foo" => 'bar', // applied - checkpoint
```
Comment (eeree) [inline src/test/resources/files/Rule/CleanCode/DuplicatedArrayKey/testRuleAppliesWhenKeyIsDeclaredInNonStandardWay.php]: I've added an additional duplicated key in the end, to check if after all the checks above, this one is not skipped.

Comment (ravage84) [inline src/test/resources/files/Rule/CleanCode/DuplicatedArrayKey/testRuleAppliesWhenKeyIsDeclaredInNonStandardWay.php]: The comment itself is not clear enough. Please, try to rephrase it.

Comment (ravage84) [inline src/test/resources/files/Rule/CleanCode/DuplicatedArrayKey/testRuleAppliesWhenKeyIsDeclaredInNonStandardWay.php]: The comment itself is not clear enough. Please, try to rephrase it.

**Comment by eeree** [comment]
@ravage84 
https://github.com/phpmd/phpmd/pull/484#pullrequestreview-42185658 done

File: src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php [hunk 3]
```diff
@@ -79,14 +79,19 @@ private function analyzeArray(ASTNode $node)
     }
 
     /**
-     * Sets normalized name as node's image.
+     * To compare keys, we have to cast them to string.
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php]: There is no short description. It should always have a short description and if necessary a long description.

https://www.phpdoc.org/docs/latest/getting-started/your-first-set-of-documentation.html#example

File: src/site/rst/rules/cleancode.rst [hunk 4]
```diff
@@ -91,10 +91,29 @@ Example: ::
...
+  }
+
+
 Remark
 ======
 
   This document is based on a ruleset xml-file, that was taken from the original source of the `PMD`__ project. This means that most parts of the content on this page are the intellectual work of the PMD community and its contributors and not of the PHPMD project.
 
 __ http://pmd.sourceforge.net/
-
```
Comment (ravage84) [inline src/site/rst/rules/cleancode.rst]: Done in https://github.com/phpmd/phpmd/commit/d40a0f1703d0bbb07b479e3a3e37750f6f5a5011

**Comment by ravage84** [comment]
🍸

**Comment by cweiske** [comment]
Since this feature has not been released 1.5 years later: [phpstan](https://github.com/phpstan/phpstan) is also able to detect duplicate array keys.

**Pull request merged**